File: 2025/python/solve_day7_part2.py
def solve_day7_part2(text):
    nrows = len(text)
    ncols = len(text[0])

    source = text[0].index("S")
    beams = [[0] * ncols for _ in range(nrows)]
    beams[0][source] = 1

    for row in range(nrows-1):
        next_row = row + 1
        beam_cols = [i for i, val in enumerate(beams[row]) if val > 0]

        for col in beam_cols:
            n_beams = beams[row][col]
            if text[next_row][col] == "^":
                beams[next_row][col+1] += n_beams
                beams[next_row][col-1] += n_beams
                continue

            beams[next_row][col] += n_beams

    return(sum(beams[nrows-1]))

File: 2025/R/solve_day7_part2.R
solve_day7_part2 <- function(input) {
  grid <- aoc_input_to_chr_matrix(input)
  nrows <- NROW(grid)
  ncols <- NCOL(grid)

  source <- which(grid == "S", arr.ind = TRUE)

  row <- source[[1L]] + 1L
  beams <- matrix(0, nrows, ncols)
  beams[row, source[[2L]]] <- 1L
  split_count <- 0L

  while(row < nrows) {
    next_row <- row + 1L
    cols <- which(beams[row, ] > 0)

    for (col in cols) {
      n_beams <- beams[row, col]
      if (grid[next_row, col] == "^") {
        split_count <- split_count + n_beams
        beams[next_row, col - 1L] <- beams[next_row, col - 1L] + n_beams
        beams[next_row, col + 1L] <- beams[next_row, col + 1L] + n_beams
        next
      }

      beams[next_row, col] <- beams[next_row, col] + n_beams
    }

    row <- next_row
  }

  sum(beams[nrows, ])
}
